### ISSUE TITLE
date filter year guessing problem

### ISSUE THREAD

**Issue by christofhaerens** [elastic/logstash#4637]
I noticed that some logs where getting the wrong timestamp in elastic via logstash.
We run logstash 2.1.1. Our setup includes logstash-fwd -> logstash input lumberjack -> logstash output redis -> redis -> logstash input redis -> grok+date filter -> logstash output elastic -> elastic
Syslog messages are in the format "Month day HH:MM:SS" and don't include the year. Because of logstash-fwd that will send events on startup since it was last active and also because of our queue, some events of dec 2015 arrive on the logstash date filter on jan 2016. But since the message has no year in it, logstash decides to timestamp it as dec 2016.

I expect that if events come in without a year that logstash could check if the month+day is in the future or not to add the current year or previous year to it

```
% echo "Dec 31 23:58:00" | /opt/logstash/bin/logstash -e 'input { stdin { } } filter { date { match => [ "message", "MMM dd HH:mm:ss" ] add_tag => [ "dated" ] } }'
Settings: Default filter workers: 4
Logstash startup completed
{
       "message" => "Dec 31 23:58:00",
      "@version" => "1",
    "@timestamp" => "2016-12-31T22:58:00.000Z",
          "host" => "bender",
          "tags" => [
        [0] "dated"
    ]
}
Logstash shutdown completed
```

**Comment by wiibaa** [comment]
Logstash 2.1.1 comest with date filter 2.0.2 => https://github.com/elastic/logstash/blob/v2.1.1/Gemfile.jruby-1.9.lock#L193

Year handling was improved  in date filter 2.1.0 => https://github.com/logstash-plugins/logstash-filter-date/pull/4#issuecomment-166392013

Can you try to update this specific plugin for your installation and see if it behave as expected for you.

**Comment by christofhaerens** [comment]
Hey, I was some days off the grid due to holiday. I will update to the latest date filter and test again. Will post the results after my tests.
Thanks

**Comment by christofhaerens** [comment]
Hi,

Seems my logstash is has already => 2.1.0:
# bin/plugin list --verbose logstash-filter-date
logstash-filter-date (2.1.1)

Also tried to do a plugin update. I'm on an internal network and set http_proxy, but I keep getting
# bin/plugin update logstash-filter-date
ERROR: Updated Aborted, message: Connection refused - Connection refused

Tried the --verbose , but seems that's no a valid option for the update cmd.

**Comment by wiibaa** [comment]
Can you check your environment locale definition, maybe you are hitting the following bug https://github.com/logstash-plugins/logstash-filter-date/issues/57

**Comment by christofhaerens** [comment]
```
# locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

**Comment by wiibaa** [comment]
Sorry for the misleading pointer, in fact it is the current behaviour of the plugin from the changelog

`Similar for if local time is January and Event time is December, the year will be set to the previous year.`
As we are now in Feb, this case does not apply.

Can you please have a look into https://github.com/logstash-plugins/logstash-filter-date/issues/51 to see if it describes your use case correctly and close this one to avoid duplicates

**Comment by christofhaerens** [comment]
Yes my example code is from Feb but we actually had the problem in Jan. I tried to reproduce on a test system and setting the date back, but then it behaved as expected.
So started digging some more and seems the prod machines had logstash 2.1.1 and my test logstash 2.1.2.
So retested with 2.1.1 and there I have the problem, which seems solved in the 2.1.2.
So yea minor version difference here that solved the problem.
Will update the prod logstash server to 2.1.2 to fix the issue.

Thxs for the help. I will close this one. That issue 51 idd describes the guessing logic relevant to my point. Next time I will create an issue on the plugin project page.

Cheers

**Comment by rblazek** [comment]
Excuse my wondering: Is the default behavior documented? I cannot find that info anywhere. Do I need to read the source code?

Thanks
Rudy
